zkir/Dialect/Field/Conversions/FieldToModArith/ExtensionFieldCodeGen.h

```diff
@@ -67,18 +141,27 @@ class ExtensionFieldCodeGen
   PrimeFieldCodeGen NonResidue() const {
     return PrimeFieldCodeGen(b, nonResidue);
   }
+  zk_dtypes::ExtensionFieldMulAlgorithm GetMulAlgorithm() const {
+    if constexpr (N == 4) {
+      return zk_dtypes::ExtensionFieldMulAlgorithm::kToomCook;
+    } else {
+      return zk_dtypes::ExtensionFieldMulAlgorithm::kKaratsuba;
+    }
+  }
   zk_dtypes::ExtensionFieldMulAlgorithm GetSquareAlgorithm() const {
-    return zk_dtypes::ExtensionFieldMulAlgorithm::kKaratsuba;
+    // Use the same algorithm as GetMulAlgorithm(). Optimizations like kCustom
+    // or kCustom2 require runtime info (limb count, non-residue value) that
+    // we don't have at codegen time. Let the canonicalizer handle it.
```

Comment:
We actually have information, you can obtain this from `value` i guess. I think based on this info, maybe you can choose `kCustom` algo!